app/demo-geant-integration/HitRootIO.cc

```diff
@@ -0,0 +1,208 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file demo-geant-integration/HitRootIO.cc
+//---------------------------------------------------------------------------//
+#include "HitRootIO.hh"
+
+#include <regex>
+#include <G4AutoDelete.hh>
+#include <G4Event.hh>
+#include <G4RunManager.hh>
+#include <G4Threading.hh>
+#include <TBranch.h>
+#include <TFile.h>
+#include <TObject.h>
+#include <TSystem.h>
+#include <TTree.h>
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/io/Logger.hh"
+#include "accel/ExceptionConverter.hh"
+#include "accel/SetupOptions.hh"
+
+#include "GlobalSetup.hh"
+#include "SensitiveHit.hh"
+
+namespace
+{
+// Mutex within the HitRootIO scope
+G4Mutex HitRootIOMutex = G4MUTEX_INITIALIZER;
+}  // namespace
+
+namespace demo_geant
+{
+G4ThreadLocal HitRootIO* HitRootIO::instance_ = nullptr;
+
+//---------------------------------------------------------------------------//
+/*!
+ * Create a ROOT output file for each worker except the master thread if MT
+ */
+HitRootIO::HitRootIO()
+{
+    G4AutoLock lock(&HitRootIOMutex);
```

Comment:
Which share resource does this lock protects? (i.e. do you really need the lock here?)

Reply:
This lock is intended to protect TFile::Open - is TFile::Open thread-safe?

Reply:
`TFile::Open` is thread safe.  It is its results (the `TFile` object can only be used in one thread at a time).

Reply:
Without the mutex, I have a crash with "`Fatal in <UnknownClass::GetListOfTypes>: fInterpreter not initialized`" which is rooted from `TFile::Open`.

Reply:
humm ... Maybe ROOT was not put in thread safe mode by calling:
```
ROOT::EnableThreadSafety();
```

Reply:
Okay.  This helps.  Added this and removed the use of G4Mutex.